test/shell/test_unused_dependency.sh

```diff
@@ -47,8 +47,41 @@ test_unused_dependency_fails_even_if_also_exists_in_plus_one_deps() {
   action_should_fail build --extra_toolchains="//test_expect_failure/plus_one_deps:plus_one_deps_with_unused_error" //test_expect_failure/plus_one_deps/with_unused_deps:a
 }
 
+test_plus_one_ast_analyzer_unused_deps() {
+  action_should_fail build --extra_toolchains="//test/toolchains:ast_plus_one_deps_unused_deps_error" //test_expect_failure/plus_one_deps/with_unused_deps:a
+  action_should_fail build --extra_toolchains="//test/toolchains:ast_plus_one_deps_strict_deps_unused_deps_error" //test_expect_failure/plus_one_deps/with_unused_deps:a
+  action_should_fail build --extra_toolchains="//test/toolchains:ast_plus_one_deps_unused_deps_error" //test_expect_failure/plus_one_deps/with_unused_deps:a
+
+
+  output=$(bazel build --extra_toolchains="//test/toolchains:ast_plus_one_deps_unused_deps_error" //test_expect_failure/plus_one_deps/with_unused_deps:a 2>&1
+  )
+
+  if [ $? -ne 0 ]; then
+    echo "Target with unused dependency failed to build with status $?"
+    echo "$output"
+    exit 1
+  fi
+
+  local expected="warning: Target '//test_expect_failure/plus_one_deps/with_unused_deps:c' is specified as a dependency to //test_expect_failure/plus_one_deps/with_unused_deps:a but isn't used, please remove it from the deps."
+
+  echo "$output" | grep "$expected"
+  if [ $? -ne 0 ]; then
+    echo "Expected output:[$output] to contain [$expected]"
+    exit 1
+  fi
```

Comment:
this is needed because `action_should_fail` doesn't allow to assert on the output? Maybe it should? If not maybe it should be extracted to a private method? It was a bit hard to understand what's going on (and maybe indeed I misunderstood)

Reply:
Ok so this was supposed to be for warn. So it's puzzling that the test passed and I need to look into that.

The logic was copied from test_unused_dependency_checker_mode_warn and therefore it is 100%  appropriate that we don't need to do anything like pull it into a function or something. (Okay the real reason is these days I do all my utility scripts in python and am terrified about bash. But will try to see if we can indeed extract a function)

Reply:
Ok so action_should_fail does `exit 0` on success so only the first test was run -_-. Not sure why we want that but will separate these tests out.

Luckily test_expect_failure_or_warning_on_missing_direct_deps_with_expected_message does not `exit 0` on success so we don't have to copy the messages around into separate functions. Though admittedly if the policy is to do an `exit 0` then its kind of risky.

Reply:
I missed the exit 0 part. If you can reiterate that would be great.
Also two more small points:
1. Do you also need to split the tests in test_deps.sh?
2. I’m glad you’re terrified from bash, it’s awful :) In general please make sure to triple check every test you’ve added/changed fails and if you change it refactor the smallest thing please make sure it still fails by changing/commenting out the prod code. We’ve had multiple issues in the past with bash tests silently failing...

Reply:
What do you mean by reiterate?

1. No because the command we run there doesn't call exit 0
2. Ok will do so in the future. If the situation is so bad wouldn't even just converting all existing tests to something where at least you won't be constantly surprised be a good idea? e.g. python or something.

Reply:
We don’t change them that often so the price is usually contained.
Also what we’d really like to do is move these tests inside to Bazel (using https://github.com/bazelbuild/bazel-integration-testing) but that’s a complicated task